# **CLEANING CSV: bank_loan_dataset_raw**

## **Load + View Dataset**

In [39]:
# Import libraries
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_csv("bank_loan_dataset_raw.csv")

In [40]:
# Inspect dataset
print("Dataset Dimensions: ", df.shape)
print('Dataset Attributes: ', df.columns.values)

df.info()
df.head(10)

Dataset Dimensions:  (100514, 19)
Dataset Attributes:  ['Loan ID' 'Customer ID' 'Loan Status' 'Current Loan Amount' 'Term'
 'Credit Score' 'Annual Income' 'Years in current job' 'Home Ownership'
 'Purpose' 'Monthly Debt' 'Years of Credit History'
 'Months since last delinquent' 'Number of Open Accounts'
 'Number of Credit Problems' 'Current Credit Balance'
 'Maximum Open Credit' 'Bankruptcies' 'Tax Liens']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100514 entries, 0 to 100513
Data columns (total 19 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Loan ID                       100000 non-null  object 
 1   Customer ID                   100000 non-null  object 
 2   Loan Status                   100000 non-null  object 
 3   Current Loan Amount           100000 non-null  float64
 4   Term                          100000 non-null  object 
 5   Credit Score                  80846 non-null   float64

Loan ID                           Customer ID  \
0  14dd8831-6af5-400b-83ec-68e61888a048  981165ec-3274-42f5-a3b4-d104041a9ca9   
1  4771cc26-131a-45db-b5aa-537ea4ba5342  2de017a3-2e01-49cb-a581-08169e83be29   
2  4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  5efb2b2b-bf11-4dfd-a572-3761a2694725   
3  77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  e777faab-98ae-45af-9a86-7ce5b33b1011   
4  d4062e70-befa-4995-8643-a0de73938182  81536ad9-5ccf-4eb8-befb-47a4d608658e   
5  89d8cb0c-e5c2-4f54-b056-48a645c543dd  4ffe99d3-7f2a-44db-afc1-40943f1f9750   
6  273581de-85d8-4332-81a5-19b04ce68666  90a75dde-34d5-419c-90dc-1e58b04b3e35   
7  db0dc6e1-77ee-4826-acca-772f9039e1c7  018973c9-e316-4956-b363-67e134fb0931   
8  8af915d9-9e91-44a0-b5a2-564a45c12089  af534dea-d27e-4fd6-9de8-efaa52a78ec0   
9  0b1c4e3d-bd97-45ce-9622-22732fcdc9a0  235c4a43-dadf-483d-aa44-9d6d77ae4583   

   Loan Status  Current Loan Amount        Term  Credit Score  Annual Income  \
0   Fully Paid             445412.0  Short Term         709.0      1167493.0   
1   Fully Paid             262328.0  Short Term           NaN            NaN   
2   Fully Paid           99999999.0  Short Term         741.0      2231892.0   
3   Fully Paid             347666.0   Long Term         721.0       806949.0   
4   Fully Paid             176220.0  Short Term           NaN            NaN   
5  Charged Off             206602.0  Short Term        7290.0       896857.0   
6   Fully Paid             217646.0  Short Term         730.0      1184194.0   
7  Charged Off             648714.0   Long Term           NaN            NaN   
8   Fully Paid             548746.0  Short Term         678.0      2559110.0   
9   Fully Paid             215952.0  Short Term         739.0      1454735.0   

  Years in current job Home Ownership             Purpose  Monthly Debt  \
0              8 years  Home Mortgage   Home Improvements       5214.74   
1            10+ years  Home Mortgage  Debt Consolidation      33295.98   
2              8 years       Own Home  Debt Consolidation      29200.53   
3              3 years       Own Home  Debt Consolidation       8741.90   
4              5 years           Rent  Debt Consolidation      20639.70   
5            10+ years  Home Mortgage  Debt Consolidation      16367.74   
6             < 1 year  Home Mortgage  Debt Consolidation      10855.08   
7             < 1 year  Home Mortgage           Buy House      14806.13   
8              2 years           Rent  Debt Consolidation      18660.28   
9             < 1 year           Rent  Debt Consolidation      39277.75   

   Years of Credit History  Months since last delinquent  \
0                     17.2                           NaN   
1                     21.1                           8.0   
2                     14.9                          29.0   
3                     12.0                           NaN   
4                      6.1                           NaN   
5                     17.3                           NaN   
6                     19.6                          10.0   
7                      8.2                           8.0   
8                     22.6                          33.0   
9                     13.9                           NaN   

   Number of Open Accounts  Number of Credit Problems  Current Credit Balance  \
0                      6.0                        1.0                228190.0   
1                     35.0                        0.0                229976.0   
2                     18.0                        1.0                297996.0   
3                      9.0                        0.0                256329.0   
4                     15.0                        0.0                253460.0   
5                      6.0                        0.0                215308.0   
6                     13.0                        1.0                122170.0   
7                     15.0                        0.0                193306.0   
8                      4.0                 

## **ISSUE #1: Fix Column Names + Data Types**
### Convert to snake_case to improve consistency and readability:
* Lowercase
* Remove leading/trailing whitespace
* Remove parentheses
* Replace dashes/spaces with underscores

In [41]:
# Standardize column names
df.columns = (df.columns
    .str.lower()
    .str.strip()
    .str.replace(" ", "_")
    .str.replace("-", "_")
    .str.replace("(", "")
    .str.replace(")", "")
)

# Inspect columns
df.columns

Index(['loan_id', 'customer_id', 'loan_status', 'current_loan_amount', 'term',
       'credit_score', 'annual_income', 'years_in_current_job',
       'home_ownership', 'purpose', 'monthly_debt', 'years_of_credit_history',
       'months_since_last_delinquent', 'number_of_open_accounts',
       'number_of_credit_problems', 'current_credit_balance',
       'maximum_open_credit', 'bankruptcies', 'tax_liens'],
      dtype='object')

## **ISSUE #2: Fix `NULL` Values**
### **`NULL`** values in this file were either left blank, or written as **`n/a`** or **`NA`**. Convert them all to **`NaN`** values, to simplify querying and visualizations.

In [44]:
# Standardize all null values
df = df.replace(
    to_replace=[
        "", " ", "  ", "   ",
        "n/a", "N/A", "na", "NA", "Na", "nA"
    ],
    value=np.nan
)

## **ISSUE #3: Fix **`credit_score`** Values**
### Some values in the **`credit_score`** column have an extra "0" at the end of the number (EX: **`credit_score`** of **`729`** is mistakenly recorded as **`7290`**. Remove this extra "0" to ensure accurate data.

In [45]:
# Fix values in credit_score column with an extra "0"
df["credit_score"] = df["credit_score"].apply(
    lambda x: x / 10 if pd.notnull(x) and x > 850 else x
)

# Make sure all credit_score values are between 300 and 850
print(df["credit_score"].describe())

count    80846.000000
mean       716.293447
std         28.297164
min        585.000000
25%        703.000000
50%        722.000000
75%        738.000000
max        751.000000
Name: credit_score, dtype: float64


## **ISSUE #4: Fix `home_ownership` Values**
### Some values in the **`home_ownership`** column record the customer's status as **`HaveMortgage`** or **`Home Mortgage`**, which is the same status written in two different ways. Convert all records of **`HaveMortgage`** into **`Home Mortgage`**, to ensure accurate data.

In [46]:
# Replace incorrect status with correct status
df["home_ownership"] = df["home_ownership"].replace(
    "HaveMortgage",
    "Home Mortgage"
)

# Make sure all incorrect values have been converted
print(df["home_ownership"].value_counts(dropna=False))

home_ownership
Home Mortgage    48624
Rent             42194
Own Home          9182
NaN                514
Name: count, dtype: int64


## **ISSUE #5: Fix `purpose` Values**
### Maintain consistency within **`purpose`** categories, to simplify querying and visualizations:
* **`moving`**, **`other`**, **`vacation`**, **`wedding`** → Title case
* **`major_purchase`**, **`small_business`**, **`renewable_energy`** → Title case, replace underscore with space
* **`Buy a Car`**, **`Buy House`**, **`Take a Trip`** → Simplify into one-word labels

In [47]:
# Standardize categories
df["purpose"] = (
    df["purpose"]
    .str.strip()
    .str.replace("_", " ", regex=False)
    .str.title()
)

# Simplify categories
purpose_simplify = {
    "Buy A Car": "Car",
    "Buy House": "House",
    "Take A Trip": "Trip"
}

df["purpose"] = df["purpose"].replace(purpose_simplify)

# Inspect categories
print(df["purpose"].value_counts(dropna=False))

purpose
Debt Consolidation      78552
Other                    9287
Home Improvements        5839
Business Loan            1569
Car                      1265
Medical Bills            1127
House                     678
Trip                      573
NaN                       514
Major Purchase            352
Small Business            283
Moving                    150
Wedding                   115
Vacation                  101
Educational Expenses       99
Renewable Energy           10
Name: count, dtype: int64


## **ISSUE #6: Feature Engineer `credit_score_range` Column**
### Create a new column that reads each value in the **`credit_score`** column and labels its FICO score category in a **`credit_score_range`** column, to enhance querying and visualizations:
* **Exceptional:** 800 - 850
* **Very Good:** 740 - 799
* **Good:** 670 - 739
* **Fair:** 580 - 669
* **Poor:** 300 - 579 

In [48]:
# Define the FICO score ranges and its associated label
conditions = [
    df["credit_score"].between(800, 850),
    df["credit_score"].between(740, 799),
    df["credit_score"].between(670, 739),
    df["credit_score"].between(580, 669),
    df["credit_score"].between(300, 579)
]

labels = [
    "Excellent",
    "Very Good",
    "Good",
    "Fair",
    "Poor"
]

# Create the new credit_score_range column
credit_score_range = np.select(conditions, labels, default="")

# Insert credit_score_range column directly after credit_score column
credit_score_index = df.columns.get_loc("credit_score") + 1
df.insert(credit_score_index, "credit_score_range", credit_score_range)

# Inspect credit_score_range column
print(df["credit_score_range"].value_counts(dropna=False))

credit_score_range
Good         56352
             19668
Very Good    18479
Fair          6015
Name: count, dtype: int64


## **View Cleaned Dataset**

In [49]:
# Inspect dataset
print("Dataset Dimensions: ", df.shape)
print('Dataset Attributes: ', df.columns.values)

df.info()
df.head(10)

Dataset Dimensions:  (100514, 20)
Dataset Attributes:  ['loan_id' 'customer_id' 'loan_status' 'current_loan_amount' 'term'
 'credit_score' 'credit_score_range' 'annual_income'
 'years_in_current_job' 'home_ownership' 'purpose' 'monthly_debt'
 'years_of_credit_history' 'months_since_last_delinquent'
 'number_of_open_accounts' 'number_of_credit_problems'
 'current_credit_balance' 'maximum_open_credit' 'bankruptcies' 'tax_liens']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100514 entries, 0 to 100513
Data columns (total 20 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   loan_id                       100000 non-null  object 
 1   customer_id                   100000 non-null  object 
 2   loan_status                   100000 non-null  object 
 3   current_loan_amount           100000 non-null  float64
 4   term                          100000 non-null  object 
 5   credit_score                  808

loan_id                           customer_id  \
0  14dd8831-6af5-400b-83ec-68e61888a048  981165ec-3274-42f5-a3b4-d104041a9ca9   
1  4771cc26-131a-45db-b5aa-537ea4ba5342  2de017a3-2e01-49cb-a581-08169e83be29   
2  4eed4e6a-aa2f-4c91-8651-ce984ee8fb26  5efb2b2b-bf11-4dfd-a572-3761a2694725   
3  77598f7b-32e7-4e3b-a6e5-06ba0d98fe8a  e777faab-98ae-45af-9a86-7ce5b33b1011   
4  d4062e70-befa-4995-8643-a0de73938182  81536ad9-5ccf-4eb8-befb-47a4d608658e   
5  89d8cb0c-e5c2-4f54-b056-48a645c543dd  4ffe99d3-7f2a-44db-afc1-40943f1f9750   
6  273581de-85d8-4332-81a5-19b04ce68666  90a75dde-34d5-419c-90dc-1e58b04b3e35   
7  db0dc6e1-77ee-4826-acca-772f9039e1c7  018973c9-e316-4956-b363-67e134fb0931   
8  8af915d9-9e91-44a0-b5a2-564a45c12089  af534dea-d27e-4fd6-9de8-efaa52a78ec0   
9  0b1c4e3d-bd97-45ce-9622-22732fcdc9a0  235c4a43-dadf-483d-aa44-9d6d77ae4583   

   loan_status  current_loan_amount        term  credit_score  \
0   Fully Paid             445412.0  Short Term         709.0   
1   Fully Paid             262328.0  Short Term           NaN   
2   Fully Paid           99999999.0  Short Term         741.0   
3   Fully Paid             347666.0   Long Term         721.0   
4   Fully Paid             176220.0  Short Term           NaN   
5  Charged Off             206602.0  Short Term         729.0   
6   Fully Paid             217646.0  Short Term         730.0   
7  Charged Off             648714.0   Long Term           NaN   
8   Fully Paid             548746.0  Short Term         678.0   
9   Fully Paid             215952.0  Short Term         739.0   

  credit_score_range  annual_income years_in_current_job home_ownership  \
0               Good      1167493.0              8 years  Home Mortgage   
1                               NaN            10+ years  Home Mortgage   
2          Very Good      2231892.0              8 years       Own Home   
3               Good       806949.0              3 years       Own Home   
4                               NaN              5 years           Rent   
5               Good       896857.0            10+ years  Home Mortgage   
6               Good      1184194.0             < 1 year  Home Mortgage   
7                               NaN             < 1 year  Home Mortgage   
8               Good      2559110.0              2 years           Rent   
9               Good      1454735.0             < 1 year           Rent   

              purpose  monthly_debt  years_of_credit_history  \
0   Home Improvements       5214.74                     17.2   
1  Debt Consolidation      33295.98                     21.1   
2  Debt Consolidation      29200.53                     14.9   
3  Debt Consolidation       8741.90                     12.0   
4  Debt Consolidation      20639.70                      6.1   
5  Debt Consolidation      16367.74                     17.3   
6  Debt Consolidation      10855.08                     19.6   
7               House      14806.13                      8.2   
8  Debt Consolidation      18660.28                     22.6   
9  Debt Consolidation      39277.75                     13.9   

   months_since_last_delinquent  number_of_open_accounts  \
0                           NaN                      6.0   
1                           8.0                     35.0   
2                          29.0                     18.0   
3                           NaN                      9.0   
4                           NaN                     15.0   
5                           NaN                      6.0   
6                          10.0                     13.0   
7                           8.0                     15.0   
8                          33.0                      4.0   
9                           NaN                     20.0   

   number_of_credit_problems  current_credit_balance  maximum_open_credit  \
0                        1.0                228190.0             416746.0   
1                        0.0                229976.0             850784.0   
2

## **Save Cleaned Dataset**

In [50]:
# Save cleaned dataset
df.to_csv("bank_loan_dataset_cleaned.csv", index=False)